src/lavinmq/http/router.cr

```diff
@@ -0,0 +1,129 @@
+require "http/server/handler"
+require "http/server/context"
+require "string_scanner"
+
+module LavinMQ::HTTP::Router
+  include ::HTTP::Handler
+
+  alias Params = Hash(String, String)
+  alias Routes = Array(Route)
+  alias Action = Proc(::HTTP::Server::Context, Params, ::HTTP::Server::Context)
+
+  abstract struct Fragment
+    def initialize(@value : String)
+    end
+
+    def match?(path : String, params : Params) : Bool
+      match?(StringScanner.new(path), params)
+    end
+
+    abstract def match?(path : StringScanner, params : Params) : Bool
+  end
+
+  struct StaticFragment < Fragment
+    FRAGMENT_PATTERN = %r{[^/]*}
+
+    def match?(path : StringScanner, params : Params) : Bool
+      path_fragment = path.scan FRAGMENT_PATTERN
+      path_fragment == @value
+    end
+  end
+
+  struct ParamFragment < Fragment
+    FRAGMENT_PATTERN = %r{[^/]*}
+
+    def match?(path : StringScanner, params : Params) : Bool
+      if path_fragment = path.scan FRAGMENT_PATTERN
+        params[@value] = path_fragment
+        return true
+      end
+      false
+    end
+  end
+
+  struct RestFragment < Fragment
+    def match?(path : StringScanner, params : Params) : Bool
+      params[@value] = path.rest
+      path.terminate
+      true
+    end
+  end
+
+  struct Fragments
+    FRAGMENT_SEPERATOR = %r{/}
+
+    def self.create(path) : Fragments
+      path = path.lstrip '/'
+      new(path.split('/').map do |value|
+        case value[0]?
+        when ':'
+          ParamFragment.new(value[1..])
+        when '*'
+          RestFragment.new(value[1..])
+        else
+          StaticFragment.new(value)
+        end
+      end)
+    end
+
+    def initialize(@fragments : Array(Fragment))
+    end
+
+    def match?(path : String, params : Params) : Bool
+      match?(StringScanner.new(path), params)
+    end
+
+    def match?(path : StringScanner, params : Params) : Bool
+      fragments = @fragments.each
+      until path.eos?
+        fragment = fragments.next
+        case fragment
+        when Iterator::Stop then return false
+        else
+          path.skip FRAGMENT_SEPERATOR
+          return false unless fragment.match?(path, params)
+        end
+      end
+      # It's a match if we've matched against all fragments
+      fragments.next == fragments.stop
+    end
+  end
+
+  struct Route
+    getter method, fragments, action
+
+    def initialize(@method : String, @path : String, @action : Action)
+      @fragments = Fragments.create(@path)
+    end
+  end
+
+  @_routes = Routes.new
+
+  {% for method in %w[delete get options patch post put] %}
```

Comment:
i guess we don't use `HEAD` but for completeness: 
```suggestion
  {% for method in %w[delete get options patch post put head] %}
```

Reply:
Actually we do (for static content), but we don't test it (it seems)!